planner/core/rule_generate_column_substitute.go

```diff
@@ -62,7 +62,7 @@ func collectGenerateColumn(lp LogicalPlan, exprToColumn ExprColumnMap) {
 			if colInfo.IsGenerated() && !colInfo.GeneratedStored {
 				s := ds.schema.Columns
 				col := expression.ColInfo2Col(s, colInfo)
-				if col != nil {
+				if col != nil && col.GetType().EvalType() == col.VirtualExpr.GetType().EvalType() {
```

Comment:
Try `col.GetType().Equal(col.VirtualExpr.GetType())`
The `EvalType` is not accurate.

Reply:
It seems `Equal` is too strict to make most of `explain_generate_column_substitute.test` pass.

Reply:
@wjhuang2016  For the case `CREATE TABLE t0(c0 CHAR AS (c1) UNIQUE, c1 INT);`. Should we treat `c0` as the `cast(c1 as char)` instead of `c1` to avoid this issue?